RecoVertex/PrimaryVertexProducer/src/DAClusterizerInZ_vect.cc

```diff
@@ -651,13 +651,16 @@ bool DAClusterizerInZ_vect::purge(vertex_t& y, track_t& tks, double& rho0, const
     double sump = 0;
 
     double pmax = y.pk_ptr[k] / (y.pk_ptr[k] + rho0 * local_exp(-beta * dzCutOff_ * dzCutOff_));
+    double p[nt];
+    for (unsigned int i = 0; i < nt; i++) {
+      p[i] = y.pk_ptr[k] * local_exp(-beta * Eik(tks.z_ptr[i], y.z_ptr[k], tks.dz2_ptr[i])) / tks.Z_sum_ptr[i];
+      //double p = tks.pi_ptr[i] * y.pk_ptr[k] * local_exp(-beta * Eik(tks.z_ptr[i], y.z_ptr[k], tks.dz2_ptr[i])) / tks.Z_sum_ptr[i];
```

Comment:
is this commented out code needed?
Please remove or add comments inline in the code why the commented out block is relevant

Reply:
Just preserving what was there before.

Reply:
@werdmann 
please clarify if this comment is essential or of it can be removed now that it falls in the context lines of modifications

Reply:
@werdmann : are tks.Z_sum_ptr[i] > 1.e-100 and tks.pi_ptr[i] > 0 just protections or intended to speed up the code?

There is no reason to introduce a double loop and the temporary array, enough to properly recast the conditions in a branchless way (of course with fast-math)

Reply:
see https://godbolt.org/z/LROrgc

Reply:
see https://godbolt.org/z/LROrgc

Reply:
the real issue with the code as is is that the compiler "think" that the protection is protecting also k (or i) not to point out of memory...

Reply:
to vectorize conditional code is good practice to copy everything in variable local to the loop scope

Reply:
or at least to force load and store inconditionally

Reply:
```
aut pp = y.pk_ptr[k] * local_exp(-beta * Eik(tks.z_ptr[i], y.z_ptr[k], tks.dz2_ptr[i])) / tks.Z_sum_ptr[i];
auto p =  (tks.Z_sum_ptr[i] > 1.e-100)  ? pp : 0;
```
will works as well and will avoid 3 more lines of code

Reply:
tkz.Z_sum_ptr[i] > 1e-100 is a protection against NaNs leading to subsequent failures. This kind of protection is present in several places in the code. I'll be happy to replace that with the better code you propose.

tks.pi_ptr[i] > 0 is not a protection but per se. And it should really be a value greater than zero to have the effect it is supposed to have, i.e. not cont tracks that are downweighted because of bad beam compatibility  in this cleaning step. That somehow slipped through the cracks but I have actually started to revisit this.

The  commented line in line 657 was an attempt pointing in the same direction. @slava77 this line can be removed.

Reply:
Thanks Wolfram for the clarification.
The code I am proposing do preserve both.

Reply:
I tried recompiling with the default compiler settings, -O2 -msse3 in a new project area and the loops don't optimize.
I added -O3 -ffast-math to package Buildfile.xml, and the compiler diagnotics indicated the loops were not optimized.
I tried adding -funroll-loops and the compiler diagnostics indicated the loops were optimized.
When I run Intel Advisor, the loops show up as not vectorized. Which do I beleive?

Reply:
I do not see how -funroll-loops  can affect vectorization heuristics (p.s. please use -Ofast)
and btw Ofast is already specified in the package BuildFile

Reply:
btw trust assembler only

Reply:
Recompiling with -Ofast did not allow the loops to vectorize.
I tried your example on the webpage with -msse3 but that did not produce vectorized loop. Using -mfma, -mavx2 or -mavx512f produces vector instructions in assemply produced. Is there an equivalent for AMD processors?

Reply:
amd understand -mfma -mavx2 and even -march=haswell
in my experience the best is to compile with -march=haswell instead of -mfma -mavx2

Reply:
in float it will vectorize with -Ofast -msse3 (maybe gcc decided that it is not worth for double?)

Reply:
ah
#pragma GCC ivdep
does the trick for sse

Reply:
btw if you try -march=wrongname of whatever gcc will be kind enough to list all allowed architecture:
```
note: valid arguments to '-march=' switch are: nocona core2 nehalem corei7 westmere sandybridge corei7-avx ivybridge core-avx-i haswell core-avx2 broadwell skylake skylake-avx512 cannonlake icelake-client icelake-server cascadelake tigerlake cooperlake bonnell atom silvermont slm goldmont goldmont-plus tremont knl knm x86-64 eden-x2 nano nano-1000 nano-2000 nano-3000 nano-x2 eden-x4 nano-x4 k8 k8-sse3 opteron opteron-sse3 athlon64 athlon64-sse3 athlon-fx amdfam10 barcelona bdver1 bdver2 bdver3 bdver4 znver1 znver2 btver1 btver2 native
``
btver1 is Naples btver2 is Rome  bdverN are various buldozer versions (Rome is the first and only AMD to implement avx2 for real)

Reply:
-Ofast -march=nehalem seems already enough to vectorize (sse4.2)